File: WebApp/src/lib/permissions.ts
export const appPermissions = {
  roles: {
    view: 'identity.roles.view',
    create: 'identity.roles.create',
    update: 'identity.roles.update',
    delete: 'identity.roles.delete',
  },
  users: {
    view: 'identity.users.view',
    create: 'identity.users.create',
    update: 'identity.users.update',
    delete: 'identity.users.delete',
  },
  categories: {
    view: 'catalog.categories.view',
    create: 'catalog.categories.create',
    update: 'catalog.categories.update',
    delete: 'catalog.categories.delete',
  },
  products: {
    view: 'catalog.products.view',
    create: 'catalog.products.create',
    update: 'catalog.products.update',
    delete: 'catalog.products.delete',
  },
  variants: {
    view: 'catalog.variants.view',
    create: 'catalog.variants.create',
    update: 'catalog.variants.update',
    delete: 'catalog.variants.delete',
  },
  customers: {
    view: 'ordering.customers.view',
    create: 'ordering.customers.create',
    update: 'ordering.customers.update',
    delete: 'ordering.customers.delete',
  },
  orders: {
    view: 'ordering.orders.view',
    create: 'ordering.orders.create',
    update: 'ordering.orders.update',
    delete: 'ordering.orders.delete',
  },
  payments: {
    view: 'payment.payments.view',
    create: 'payment.payments.create',
    update: 'payment.payments.update',
    delete: 'payment.payments.delete',
  },
} as const;

export const permissionGroups = [
  { label: 'Roles', permissions: [appPermissions.roles.view, appPermissions.roles.create, appPermissions.roles.update] },
  { label: 'Users', permissions: [appPermissions.users.view, appPermissions.users.create, appPermissions.users.update, appPermissions.users.delete] },
  { label: 'Categories', permissions: [appPermissions.categories.view, appPermissions.categories.create, appPermissions.categories.update, appPermissions.categories.delete] },
  { label: 'Products', permissions: [appPermissions.products.view, appPermissions.products.create, appPermissions.products.update, appPermissions.products.delete] },
  { label: 'Variants', permissions: [appPermissions.variants.view, appPermissions.variants.create, appPermissions.variants.update, appPermissions.variants.delete] },
  { label: 'Customers', permissions: [appPermissions.customers.view, appPermissions.customers.create, appPermissions.customers.update, appPermissions.customers.delete] },
  { label: 'Orders', permissions: [appPermissions.orders.view, appPermissions.orders.create, appPermissions.orders.update, appPermissions.orders.delete] },
  { label: 'Payments', permissions: [appPermissions.payments.view, appPermissions.payments.update] },
] as const;

const permissionActionOrder = ['view', 'create', 'update', 'delete'] as const;

export function formatPermission(permission: string) {
  const parts = permission.split('.');
  return parts.map((part) => part.charAt(0).toUpperCase() + part.slice(1)).join(' / ');
}

export function formatPermissionAction(permission: string) {
  const action = permission.split('.').at(-1) ?? permission;
  return action.charAt(0).toUpperCase() + action.slice(1);
}

export function summarizePermissions(permissions: string[]) {
  const actionsByEntity = new Map<string, Set<string>>();

  for (const permission of permissions) {
    const parts = permission.split('.');
    if (parts.length < 3) {
      continue;
    }

    const entity = parts[1];
    const action = parts[2];
    const currentActions = actionsByEntity.get(entity) ?? new Set<string>();
    currentActions.add(action);
    actionsByEntity.set(entity, currentActions);
  }

  return permissionGroups
    .map((group) => {
      const entityKey = group.permissions[0].split('.')[1];
      const actions = actionsByEntity.get(entityKey);
      if (!actions?.size) {
        return null;
      }

      const orderedActions = permissionActionOrder
        .filter((action) => actions.has(action))
        .map((action) => action.charAt(0).toUpperCase() + action.slice(1));

      return `${group.label}/${orderedActions.join('+')}`;
    })
    .filter((item): item is string => item !== null);
}

File: Shared/Authentication/Permissions.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.Extensions.DependencyInjection;

namespace Shared.Authentication;

public static class PermissionClaimTypes
{
    public const string Permission = "permission";
}

public static class ApplicationClaimTypes
{
    public const string CustomerId = "customerId";
}

public static class Permissions
{
    public static class Roles
    {
        public const string View = "identity.roles.view";
        public const string Create = "identity.roles.create";
        public const string Update = "identity.roles.update";
        public const string Delete = "identity.roles.delete";
    }

    public static class Users
    {
        public const string View = "identity.users.view";
        public const string Create = "identity.users.create";
        public const string Update = "identity.users.update";
        public const string Delete = "identity.users.delete";
    }

    public static class Categories
    {
        public const string View = "catalog.categories.view";
        public const string Create = "catalog.categories.create";
        public const string Update = "catalog.categories.update";
        public const string Delete = "catalog.categories.delete";
    }

    public static class Products
    {
        public const string View = "catalog.products.view";
        public const string Create = "catalog.products.create";
        public const string Update = "catalog.products.update";
        public const string Delete = "catalog.products.delete";
    }

    public static class Variants
    {
        public const string View = "catalog.variants.view";
        public const string Create = "catalog.variants.create";
        public const string Update = "catalog.variants.update";
        public const string Delete = "catalog.variants.delete";
    }

    public static class Customers
    {
        public const string View = "ordering.customers.view";
        public const string Create = "ordering.customers.create";
        public const string Update = "ordering.customers.update";
        public const string Delete = "ordering.customers.delete";
    }

    public static class Orders
    {
        public const string View = "ordering.orders.view";
        public const string Create = "ordering.orders.create";
        public const string Update = "ordering.orders.update";
        public const string Delete = "ordering.orders.delete";
    }

    public static class Payments
    {
        public const string View = "payment.payments.view";
        public const string Create = "payment.payments.create";
        public const string Update = "payment.payments.update";
        public const string Delete = "payment.payments.delete";
    }

    public static readonly IReadOnlyList<string> All =
    [
        Roles.View,
        Roles.Create,
        Roles.Update,
        Roles.Delete,
        Users.View,
        Users.Create,
        Users.Update,
        Users.Delete,
        Categories.View,
        Categories.Create,
        Categories.Update,
        Categories.Delete,
        Products.View,
        Products.Create,
        Products.Update,
        Products.Delete,
        Variants.View,
        Variants.Create,
        Variants.Update,
        Variants.Delete,
        Customers.View,
        Customers.Create,
        Customers.Update,
        Customers.Delete,
        Orders.View,
        Orders.Create,
        Orders.Update,
        Orders.Delete,
        Payments.View,
        Payments.Create,
        Payments.Update,
        Payments.Delete,
    ];

    public static readonly IReadOnlyDictionary<string, IReadOnlyList<string>> Groups =
        new Dictionary<string, IReadOnlyList<string>>(StringComparer.OrdinalIgnoreCase)
        {
            ["Roles"] = [Roles.View, Roles.Create, Roles.Update, Roles.Delete],
            ["Users"] = [Users.View, Users.Create, Users.Update, Users.Delete],
            ["Categories"] = [Categories.View, Categories.Create, Categories.Update, Categories.Delete],
            ["Products"] = [Products.View, Products.Create, Products.Update, Products.Delete],
            ["Variants"] = [Variants.View, Variants.Create, Variants.Update, Variants.Delete],
            ["Customers"] = [Customers.View, Customers.Create, Customers.Update, Customers.Delete],
            ["Orders"] = [Orders.View, Orders.Create, Orders.Update, Orders.Delete],
            ["Payments"] = [Payments.View, Payments.Create, Payments.Update, Payments.Delete],
        };
}

public static class AuthorizationConfigurationExtensions
{
    public static IServiceCollection AddApplicationAuthorization(this IServiceCollection services)
    {
        services.AddAuthorization(options =>
        {
            foreach (var permission in Permissions.All)
            {
                options.AddPolicy(permission, policy =>
                {
                    policy.RequireAuthenticatedUser();
                    policy.RequireAssertion(context =>
                        context.User.IsInRole("Admin") ||
                        context.User.HasClaim(PermissionClaimTypes.Permission, permission));
                });
            }
        });

        return services;
    }
}
